Enhanced Phase Change Material Simulation

Myoldmopar opened the pull request:
Pull request overview
---------------------
Still work to do here.
 - [x] Update IDD fields with descriptions, defaults, units, etc.
 - [x] Add error handling to input processor
 - [ ] Tentatively add zone level enthalpy outputs
 - [x] Remove any temporary outputs, etc.
 - [ ] Final review

Myoldmopar committed:
Initial outline of PCM enhancement design doc

Myoldmopar committed:
Comment-only clean up in the code, updating design doc

Myoldmopar committed:
Comment-only code cleanup, continued work on the design doc

Myoldmopar committed:
Update doc

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into PCM-NRGSim-NREL-SBV

Myoldmopar committed:
Rough starting to throw in some PCM code

Myoldmopar committed:
Merge develop; resolve conflict; verify resolution

Myoldmopar committed:
Fix idd conflict issue

Myoldmopar committed:
Add new generalized phase change material base class and the start of the hysteresis model class

Myoldmopar committed:
Correct IDD object, add in simple input processing

Myoldmopar committed:
Initial transition from the fortran cp calculation into the hysteresis model class

Myoldmopar committed:
Continued stubbing out model implementation

Myoldmopar committed:
Intermediate commit for merging in final 8.7.0

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into PCM-NRGSim-NREL-SBV

Myoldmopar committed:
Clean up pointer dereference

Myoldmopar committed:
Semi-running code, still plenty of debugging and looking, but it at least runs...

nrel-bot-3 commented:
@Myoldmopar @lgentile it has been 14 days since this pull request was last updated.

Myoldmopar committed:
Adding in PCM checks for the rest of the configurations

Myoldmopar committed:
Simple comment only cleanup

Myoldmopar committed:
Merge branch 'PCM-NRGSim-NREL-SBV' of github.com:NREL/EnergyPlus into PCM-NRGSim-NREL-SBV

Myoldmopar committed:
Better tracking of phase change state per node

Myoldmopar committed:
Continued working on missing pieces and debugging

Myoldmopar committed:
Intermediate commit

Myoldmopar committed:
Merge, resolve conflicts

Myoldmopar committed:
Close to functional

Myoldmopar committed:
Added reporting, fixes

Myoldmopar committed:
Cleanup

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into PCM-NRGSim-NREL-SBV

Myoldmopar committed:
Cleanup unused vars, comments

Myoldmopar committed:
Use of enums, added temp reversal, cleanup license

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into PCM-NRGSim-NREL-SBV

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into PCM-NRGSim-NREL-SBV

Myoldmopar committed:
Add initial unit test, next flesh it out!

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into PCM-NRGSim-NREL-SBV

Myoldmopar committed:
Fleshing out unit test, looking good so far

Myoldmopar committed:
Continued adding unit tests

Myoldmopar committed:
Updated unit test with proper temps; created unit test gen function

Myoldmopar committed:
Continued adding unit tests

Myoldmopar committed:
Updated IDD object, added IORef and EngRef changes

Myoldmopar committed:
Clean up input processing of hysteresis models

Myoldmopar committed:
Remove temporary Cp output

Myoldmopar commented:
@mjwitte @JasonGlazer I'm going to create a tag of this, and if you happen to have a chance sometime soon, could you test out the Windows build?  I've been unable to replicate issues that the key user on this PR is seeing.

Myoldmopar committed:
Removed inheritance, will try that again later; fixed up PCM zone enthalpy outputs

Myoldmopar committed:
Made Enthalpy report vars advanced only

Myoldmopar commented:
Addressing review comments from long ago.

@JasonGlazer I removed the second latent heat term, there is now one latent term defining the energy change from solid to melted or vice-versa.  The path to get from one state to the other is different, and allows hysteresis, but this makes much more sense, energy-wise.

@michael-okeefe I worked with Kosny in the past, and he was always interested in seeing the PCM model improved in E+ to capture more physical phenomena, which this does.  This model is indeed a hT curve, so it simply fits within the existing finite difference simulation model as a new way to calculate Cp.  This is implemented as an alternative modeling approach, so that users can input data based on the amount of input data they have:

- No PCM data: just use constant properties
- Some critical parameters: use this new model and it will create an hT curve for you
- Full experimental test data: use the tabular input for PCM data to define the detailed hT curve

This could be expanded in the future, I could see adding in the hysteresis effect to the tabular test data input method, but that is well outside of the scope of this project.

I have received word from the SBV PI that at this point, the code is behaving as expected, and I think I am done.  I'll be looking for a reviewer to look this over before they/I merge it.  I'll ping relevant folks.

mbadams5 commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -5191,11 +5191,13 @@ CalcHeatBalanceInsideSurf( Optional_int_const ZoneToResimulate ) // if passed in
 	int OtherSideZoneNum; // Zone Number index for other side of an interzone partition HAMT
 	static int WarmupSurfTemp;
 	static int TimeStepInDay( 0 ); // time step number
+	static Array1D< Real64 > SurfaceEnthalpyRead; // ! True if EnthalpyM and EnthalpyH are read
```
Why create a new static function variable? It doesn't seem to cause unit testing issues in this PR but could in the future.

Myoldmopar replied:
Moving to namespace variable and adding to `clear_state`.

mbadams5 commented on src/EnergyPlus/PhaseChangeModeling/HysteresisModel.cc:
```diff
@@ -0,0 +1,344 @@
+	int numHysteresisModels = 0;
+	std::vector< HysteresisPhaseChange > hysteresisPhaseChangeModels;
+
+	HysteresisPhaseChange * HysteresisPhaseChange::factory( const std::string objectName ) {
```
Should pass by reference.

Myoldmopar replied:
Doing it!

mbadams5 commented on src/EnergyPlus/PhaseChangeModeling/HysteresisModel.cc:
```diff
@@ -0,0 +1,344 @@
+	Real64 HysteresisPhaseChange::getEnthalpy( Real64 T, Real64 Tc, Real64 tau1, Real64 tau2 ) {
+		// Looks up the enthalpy on the characteristic curve defined by the parameters Tc, tau1, and tau2,
+		// and the position on that curve defined by T.
+		Real64 eta1 = ( this->totalLatentHeat / 2 ) * exp( -2 * abs( T - Tc ) / tau1 );
```
Can `tau1` or `tau2` be zero? If so, might need to protect against divide by zero.

Myoldmopar replied:
tau1 and tau2 will always be assigned a meaningful value from the input object.  The input object includes \minimum>0 so these values should be guaranteed to always have meaningful, >0, values.  I am cleaning up the logic at the top of this function while addressing this.

mbadams5 commented on src/EnergyPlus/PhaseChangeModeling/HysteresisModel.cc:
```diff
@@ -0,0 +1,344 @@
+				this->enthalpyM = this->getEnthalpy(updatedTempTDT, this->peakTempMelting, this->deltaTempMeltingLow, this->deltaTempMeltingHigh );
+				this->enthalpyF = this->getEnthalpy(updatedTempTDT, this->peakTempFreezing, this->deltaTempFreezingLow, this->deltaTempFreezingHigh );
+				if ( this->enthNew < this->enthRev && this->enthNew >= this->enthalpyF && updatedTempTDT <= prevTempTD) {
+					phaseChangeState = 1;
```
Should probably be consistent and use `PhaseChangeStates::FREEZING`. The same is true for all other instances where you directly use the number value instead of the "enum".

Myoldmopar replied:
:+1:

mbadams5 commented on src/EnergyPlus/PhaseChangeModeling/HysteresisModel.cc:
```diff
@@ -0,0 +1,344 @@
+        Real64 Cp2 = this->specificHeatLiquid;
+        Real64 T = temperatureCurrent;
+
+        Real64 DEta1 = - ( this->totalLatentHeat * ( T - criticalTemperature ) * exp( -2 * abs( T - criticalTemperature ) / tau1 ) ) / ( tau1 * abs( T - criticalTemperature ) );
```
Same comment about `tau1` and `tau2` and divide by zero. Also can `abs( T - criticalTemperature )` ever be zero? It appears so due to ` } else if ( T == criticalTemperature ) {` below...

Myoldmopar replied:
Reworked that function so that the intermediate variables are only calculated when valid, good to go here now.

mbadams5 commented on src/EnergyPlus/PhaseChangeModeling/HysteresisModel.hh:
```diff
@@ -0,0 +1,125 @@
+
+		Real64 getEnthalpy( Real64 T, Real64 Tc, Real64 tau1, Real64 tau2 );
+
+		Real64 getCurrentSpecificHeat( Real64, Real64, Real64, int, int & );
```
I know you can do this, but I personally don't particularly like it without variable names.

Myoldmopar replied:
I don't like it either.  I was doing it early on when the argument list was fluid and changing a lot, and forgot to go back and add them in.  I fixed it.

mbadams5 commented on src/EnergyPlus/PhaseChangeModeling/HysteresisModel.cc:
```diff
@@ -0,0 +1,344 @@
+
+			// we just need to loop over the existing vector elements to check for duplicates since we haven't add this one yet
+			for ( auto & existingHysteresisModel : hysteresisPhaseChangeModels ) {
+				if ( DataIPShortCuts::cAlphaArgs( 1 ) == existingHysteresisModel.name ) {
```
Adding to then comparing to a `std::set`/`std::unordered_set` or something would be better here. That would change it from `O(n^2)` to `O(n log n)`/`O(n)` but i'm fine with this as is for now.

Myoldmopar replied:
OK, I'll pass on making changes for now.  Even for PCM heavy files, this will only be looping over like 5-10 objects, once during setup.

mbadams5 commented on src/EnergyPlus/PhaseChangeModeling/HysteresisModel.hh:
```diff
@@ -0,0 +1,125 @@
+
+#include <EnergyPlus.hh>
+#include <DataGlobals.hh>
+#include <OutputProcessor.hh>
```
Why do you need `OutputProcessor.hh` and `DataGlobals.hh` includes in the header file? Could these move to the implementation file? If includes are not needed in the header file it is better to only include them in the implementation file.

Myoldmopar replied:
Turns out I didn't need them, removed.

mbadams5 commented on tst/EnergyPlus/unit/PhaseChangeModeling/HysteresisModel.unit.cc:
```diff
@@ -0,0 +1,677 @@
+	Real64 expectedSpecificHeat;
+
+	GetSpecHeatArgs(Real64 _previousTemperature, Real64 _updatedTemperature, Real64 _temperatureReverse,
+					int _previousPhaseChangeState, int _expectedUpdatedPhaseChangeState, Real64 _expectedSpecificHeat) :
```
Looks like formatting issue.

Myoldmopar replied:
Took out the wrapping.

mbadams5 commented on tst/EnergyPlus/unit/PhaseChangeModeling/HysteresisModel.unit.cc:
```diff
@@ -0,0 +1,677 @@
+	//generate_test_data(this->ModelA, "StraightUpCurve", 14, 26, 0.5, 2);
+	//generate_test_data(this->ModelA, "StraightDownCurve", 30, 15, -0.5, -2);
+	//generate_test_data_two_directions(this->ModelA, "IntoMeltingAndBackDown", 14, 0.5, 20, -0.5, 14, 2);
+	generate_test_data_two_directions(this->ModelA, "CompletelyThroughMeltingAndBackDown", 14, 0.5, 30, -0.5, 14, 2);
```
Do these generate calls still need to be generated by the unit test? Also do they still need to be included with this PR?

Myoldmopar replied:
Accidental leftover, took all that stuff out.

mbadams5 commented on tst/EnergyPlus/unit/PhaseChangeModeling/HysteresisModel.unit.cc:
```diff
@@ -0,0 +1,677 @@
+class HysteresisTest : public testing::Test
+{
+public:
+	HysteresisPhaseChange::HysteresisPhaseChange ModelA;
```
I assume this is getting generated new for each unit test?

Myoldmopar replied:
On each unit test, it should be calling setUp(), right?  And setUp should make sure it is fully populated no matter which unit test is calling it.

Myoldmopar committed:
Address review comments

Myoldmopar commented:
@mbadams5 All review comments addressed via 1686368.  Unless you have a remaining issue with the unit test `setUp` stuff, this is complete.  If you are satisfied and want to merge it sometime, feel free, otherwise I'll do it after CI comes back happy.

Myoldmopar commented:
And thanks for the thorough review!

Myoldmopar committed:
Clean up indentation

Myoldmopar committed:
Make methods private in hysteresis code

Myoldmopar committed:
Code style cleanup

Myoldmopar commented:
![image](https://user-images.githubusercontent.com/849824/27617427-d779972a-5b7b-11e7-9423-632a3e1e7b39.png)
:+1: :+1: :+1:
